docs/reference/remap/del.cue

```diff
@@ -3,28 +3,28 @@ package metadata
 remap: functions: del: {
 	arguments: [
 		{
-			name:        "paths"
-			description: "The paths of the fields to delete."
+			name:        "path"
+			description: "The path of the field to delete."
 			required:    true
-			multiple:    true
 			type: ["string"]
 		},
 	]
-	return: ["null"]
+	return: ["any"]
 	category: "Event"
 	description: #"""
-		Removed the fields specified by the given paths from the root `event` object. Multiple fields can be specified.
+		Removed the field specified by the given path from the event object. If the field exists,
+		the field's value is returned by the delete operation; if the field does not exist, `null`
+		is returned.
```

Comment:
I'm not sure where I proposed this, but I proposed that we return an _error_ if the field does not exist.

That is the only way we can distinguish between a field that exists and has a `null` value, and a field that doesn't exist.

Reply:
Here's the issue @binarylogic created based on my earlier comment: https://github.com/timberio/vector/issues/5779

Reply:
We decided to just return Null for now and revisit it again once we have fully nailed down the error handling. See comment from @binarylogic above...

*(Hope you had a good holiday!)*

Reply:
> (Hope you had a good holiday!)

I had, thanks @FungusHumungus 😄 

As for the decision to return `null`, I don't think that's a good idea.

For one, it means you can't reliably use `del` to return the value of the deleted field, since you don't know if the field contained null or didn't exist. Second, we want to resolve error handling before we ship the first version regardless, so we might as well move ahead and make this function fallible while we also continue to work on the error handling story.

Reply:
So, I've seen the example given in the announcement post:

```rust
ok(del(.user.id))
ok(del(.user.ip_address))
.log_stream_name = "system_logs"
```

I've also since this comment written about our [final error handling story](https://gist.github.com/JeanMertz/f36e0aae6191231d1a37d81ba560047e). Specifically, I've written about allowing functions to [define their fallibility](https://gist.github.com/JeanMertz/f36e0aae6191231d1a37d81ba560047e#per-function-ignore-error-options) based on parameters.

Given the new context and the provided example, I've come around to my earlier conclusion.

Let's keep the above example, and go through the different possibilities:

1. Use the `!` bang-function to abort.

   ```rust
   del!(.user.id)
   del!(.user.ip_address)
   .log_stream_name = "system_logs"
   ```

   This works, but I can't imagine you want to abort the program here because if the field didn't exist, you don't mind if it wasn't deleted anyway. The end-result is the same, regardless if the `del` function call succeeded.

2. Use the `?` error coalesce operator to silence errors.

   ```rust
   del(.user.id) ? null
   del(.user.ip_address) ? null
   .log_stream_name = "system_logs"
   ```

   This looks weird, since you're not doing anything with the `null` value, so it's basically a way to hide the error.

3. Use the error-assignment technique to handle/ignore errors.

   ```rust
   _, _ = del(.user.id)
   _, _ = del(.user.ip_address)
   .log_stream_name = "system_logs"
   ```

   This has the same problem as the error coalesce operator except worse, there's a lot of ugly syntax here just to hide an error because we don't care about both the error and the return value in this case.

4. Use the `ignore_errors` parameter to silence errors.

   ```rust
   del(.user.id, true)
   del(.user.ip_address, true)
   .log_stream_name = "system_logs"
   ```

   I omitted the `ignore_errors` parameter keyword for brevity, but even so, this is more verbose than you'd like for something you'll use as frequent as the `del` function.

Given the above examples, I'm inclined to go with (4), but flip the defaults:

If the second argument is `false` (or undefined), then `del` is an infallible function that returns `null` if the field didn't exist, and otherwise returns the value of the field. If it's set to `true`, then the function becomes fallible, and errors if the field didn't exist:

```rust
del(.user.id)
del!(.user.ip_address, true) // this aborts if the field didn't exist
.log_stream_name = "system_logs"
```

As for my reasoning:

The first argument for this proposal is that it is much easier on the eyes, and (I believe) favours the most frequent use-case (don't care about the return value, or more specifically, don't care if I can't distinguish between `null` fields and non-existing fields).

The second argument (which I hadn't thought about until just now) is that it is actually _more consistent_ with other parts of the language. Specifically, if I query a field `.foo.bar`, and that field does not exist, we _also_ return `null` as the default value, so even there you can't know if the field was set and has `null` as its value, or if the field doesn't exist (we provide the `exists` function to make that distinction, which can also be used when calling `del`).

So, in conclusion, my proposal would be:

* have `del` take two arguments: `value` and `error_when_missing`
* `error_when_missing` defaults to `false`
* make `del` infallible by default, and return `null` if the field doesn't exist
* `error_when_missing = true` makes `del` become fallible, returning an error if the field doesn't exist

_(implementation detail note: the `error_when_missing` parameter has to take a _literal boolean_ as its value because we have to determine the value of the boolean at compile-time to be able to define the function as fallible or infallible based on this value)_

Reply:
> (implementation detail note: the error_when_missing parameter has to take a literal boolean as its value because we have to determine the value of the boolean at compile-time to be able to define the function as fallible or infallible based on this value)

This caveat is a bit unfortunate. I realize why and that users are unlikely to want to use a variable in this particular case, but it still makes the language feel a bit inconsistent to me. If we really want to pursue this route (see my alternative suggestion below) than I might advocate for making this two separate functions: `del` and `edel` (we can work on the names) to avoid introducing this special case.

I think I'm actually not a fan of having `del()` change "abort" behavior based on a boolean flag as I think this is slightly confusing as a user and complicates the internal implementation by requiring the above caveat about it being a literal boolean. I might again prefer to borrow a pattern from Go and have it simply return two values:

* The value of the deleted path or `null` if the path was undefined
* A boolean indicating whether there was actually anything at that path

Users would call this as:

```
user_id, ok = del(.user.id)
```

They can then check `ok` to see if `user_id` was actually defined.

Alternatively we simply keep it as `user_id = del(.user_id)` and users can check if `.user_id` was defined before calling `del()` if they care (pseudocode):

```
if (is_defined(.user_id)) {
  user_id = del(.user_id)
} else {
  user_id = "unknown"
}
```

Notably if we don't change `del()` to return the value, users can also simply do:

```
user_id = .user_id
del(.user_id)
```

Which I don't think is that bad for the initial release and we could modify `del()`'s behavior later in a compatible fashion.

Reply:
Thanks for the input @jszwedko. I think we diverge a bit on the priorities for the language (which is on me, since I haven't taken the time to flesh out the specification yet in https://github.com/timberio/vector/issues/3740).

I think your proposal veers too much from what we've been doing so far with the language (using separate functions for error/non-error and using multiple return values), but obviously, we can pivot into such a direction if others agree.

Of the two proposals, I'd favour separate functions over multiple return values, but I still favour adding the proposed argument to `del`.

To add a few more points supporting that proposal:

* we already have literal string and array requirements for some arguments, so I don't consider literal booleans to be a problem
* the goal of the proposal/this PR is to make `del` as easy to use as possible, while still allowing "accurate" usage if needed, so while we already support `exists` and if-statements, I don't consider those to be valid alternatives
* in the end, I think we should optimize for the 80% use-case, while providing a low-key solution for the other 20%, which is why I prefer infallible `del` by default, and a flag to make it fallible (but more correct).

I want to avoid this devolving into a lengthy discussion, so if you want, we can discuss this on Discord further to reach consensus.

Reply:
@jszwedko and I discussed this offline. It's always a pleasure to debate things like this with you, Jesse 😄.

Here's the conclusion:

I misread this part of Jesse's comment:

> Alternatively we simply keep it as `user_id = del(.user_id)` and users can check if `.user_id` was defined before calling `del()` if they care (pseudocode):
> 
> ```rust
> if exists(.user.id) {
>     .user_id = del(.user.id)
> } else {
>     .user_id = "unknown"
> }
> ```

Having now understood this part of the proposal better, given the fact that we want to "optimize for the 80% use-case", I agree with this direction.

So in conclusion, what we want to do is:

1. make `del` always infallible
2. have it return the value of the deleted field
3. return `null` if a field doesn't exist
4. document the usage of `exists` for people that want to distinguish between existing `null`, and non-existing fields (minority use-case)